File: lib/features/orders/infra/repositories/order_repository_impl.dart
import 'package:result_dart/result_dart.dart';

import '../../../../core/errors/error_handler.dart';
import '../../domain/entities/order.dart';
import '../../domain/errors/failures.dart';
import '../../domain/repositories/order_repository.dart';
import '../datasources/order_datasource.dart';

class OrderRepositoryImpl implements IOrderRepository {
  final IErrorHandler _errorHandler;
  final IOrderDatasource _orderDatasource;

  OrderRepositoryImpl(this._errorHandler, this._orderDatasource);

  @override
  AsyncResult<Order, OrdersFailure> getOrderByTypeAndNumber(
    String type,
    String number,
  ) async {
    try {
      final order = await _orderDatasource.getOrderByTypeAndNumber(
        type,
        number,
      );

      return Success(order);
    } on OrdersFailure catch (failure) {
      return Failure(failure);
    } catch (exception, stackTrace) {
      await _errorHandler.reportException(exception, stackTrace);
      return const Failure(UnknownError());
    }
  }

  @override
  AsyncResult<List<Order>, OrdersFailure> getAllOrdersBySendDate(
    String sendDate,
  ) async {
    try {
      final orders = await _orderDatasource.getAllOrdersBySendDate(sendDate);

      return Success(orders);
    } on OrdersFailure catch (failure) {
      return Failure(failure);
    } catch (exception, stackTrace) {
      await _errorHandler.reportException(exception, stackTrace);
      return const Failure(UnknownError());
    }
  }

  @override
  AsyncResult<List<Order>, OrdersFailure> getAllOrdersByArrivalDate(
    String arrivalDate,
  ) async {
    try {
      final orders = await _orderDatasource.getAllOrdersByArrivalDate(
        arrivalDate,
      );

      return Success(orders);
    } on OrdersFailure catch (failure) {
      return Failure(failure);
    } catch (exception, stackTrace) {
      await _errorHandler.reportException(exception, stackTrace);
      return const Failure(UnknownError());
    }
  }

  @override
  AsyncResult<Order, OrdersFailure> saveOrder(Order order) async {
    try {
      final savedOrder = await _orderDatasource.saveOrder(order);

      return Success(savedOrder);
    } on OrdersFailure catch (failure) {
      return Failure(failure);
    } catch (exception, stackTrace) {
      await _errorHandler.reportException(exception, stackTrace);
      return const Failure(UnknownError());
    }
  }

  @override
  AsyncResult<Unit, OrdersFailure> deleteOrder(Order order) async {
    try {
      await _orderDatasource.deleteOrder(order);

      return const Success(unit);
    } on OrdersFailure catch (failure) {
      return Failure(failure);
    } catch (exception, stackTrace) {
      await _errorHandler.reportException(exception, stackTrace);
      return const Failure(UnknownError());
    }
  }
}

File: src/features/orders/repositories/order-repository-impl.ts
import { OrderRepository } from 'src/features/orders/repositories/order-repository';
import { Order } from 'src/features/orders/types/order';
import { OrderType } from 'src/features/orders/types/order-type';
import {
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  getFirestore,
  query,
  setDoc,
  where,
} from 'firebase/firestore';
import { OrderNotFound } from 'src/features/orders/errors/order-failures';
import { orderMapper } from 'src/features/orders/repositories/mappers/order-mapper';

export const OrderRepositoryImpl = (): OrderRepository => ({
  /**
   * Retrieve from Firestore an order by type and number.
   *
   * @param type
   * @param number
   * @throws OrderNotFound
   */
  async getOrderByTypeAndNumber(type: OrderType, number: string): Promise<Order> {
    const firestore = getFirestore();
    const orderId = `${type}_${number}`;
    const ordersCollection = collection(firestore, 'orders');
    const orderDoc = doc(ordersCollection, orderId);
    const orderSnapshot = await getDoc(orderDoc);

    if (!orderSnapshot.exists()) {
      throw new OrderNotFound();
    }

    const orderData = orderSnapshot.data();

    return orderMapper.fromDocumentData(orderData);
  },

  /**
   * Retrieve from Firestore an orders list by send date.
   *
   * @param sendDate
   */
  async getAllOrdersBySendDate(sendDate: string): Promise<Order[]> {
    const firestore = getFirestore();
    const ordersCollection = collection(firestore, 'orders');
    const ordersQuery = query(ordersCollection, where('sendDate', '==', sendDate));
    const ordersSnapshot = await getDocs(ordersQuery);
    const ordersDocs = ordersSnapshot.docs;

    return ordersDocs
      .map((orderSnapshot) => orderSnapshot.data())
      .map((orderData) => orderMapper.fromDocumentData(orderData));
  },

  /**
   * Retrieve from Firestore an orders list by arrival date.
   *
   * @param arrivalDate
   */
  async getAllOrdersByArrivalDate(arrivalDate: string): Promise<Order[]> {
    const firestore = getFirestore();
    const ordersCollection = collection(firestore, 'orders');
    const ordersQuery = query(ordersCollection, where('arrivalDate', '==', arrivalDate));
    const ordersSnapshot = await getDocs(ordersQuery);
    const ordersDocs = ordersSnapshot.docs;

    return ordersDocs
      .map((orderSnapshot) => orderSnapshot.data())
      .map((orderData) => orderMapper.fromDocumentData(orderData));
  },

  /**
   * Save or update an order in Firestore.
   *
   * @param order
   */
  async saveOrder(order: Order): Promise<Order> {
    const firestore = getFirestore();
    const orderId = `${order.type}_${order.number}`;
    const ordersCollection = collection(firestore, 'orders');
    const orderDoc = doc(ordersCollection, orderId);
    const orderData = orderMapper.toDocumentData(order);

    await setDoc(orderDoc, orderData);

    return order;
  },

  /**
   * Delete an order in Firestore.
   *
   * @param order
   */
  async deleteOrder(order: Order): Promise<void> {
    const firestore = getFirestore();
    const orderId = `${order.type}_${order.number}`;
    const ordersCollection = collection(firestore, 'orders');
    const orderDoc = doc(ordersCollection, orderId);

    await deleteDoc(orderDoc);
  },
});
